compiler/natives/src/internal/reflectlite/reflectlite.go

```diff
@@ -0,0 +1,973 @@
+	outCount uint16
+
+	_in  []*rtype
+	_out []*rtype
+}
+
+func (t *funcType) in() []*rtype {
+	return t._in
+}
+
+func (t *funcType) out() []*rtype {
+	return t._out
+}
+
+type name struct {
+	bytes *byte
+}
+
+type nameData struct {
+	name     string
+	tag      string
+	exported bool
+}
+
+var nameMap = make(map[*byte]*nameData)
+
+func (n name) name() (s string) { return nameMap[n.bytes].name }
+func (n name) tag() (s string)  { return nameMap[n.bytes].tag }
+func (n name) pkgPath() string  { return "" }
+func (n name) isExported() bool { return nameMap[n.bytes].exported }
+
+func newName(n, tag string, exported bool) name {
+	b := new(byte)
+	nameMap[b] = &nameData{
+		name:     n,
+		tag:      tag,
+		exported: exported,
+	}
+	return name{
+		bytes: b,
+	}
+}
+
+var nameOffList []name
+
+func (t *rtype) nameOff(off nameOff) name {
+	return nameOffList[int(off)]
+}
+
+func newNameOff(n name) nameOff {
+	i := len(nameOffList)
+	nameOffList = append(nameOffList, n)
+	return nameOff(i)
+}
+
+var typeOffList []*rtype
+
+func (t *rtype) typeOff(off typeOff) *rtype {
+	return typeOffList[int(off)]
+}
+
+func newTypeOff(t *rtype) typeOff {
+	i := len(typeOffList)
+	typeOffList = append(typeOffList, t)
+	return typeOff(i)
+}
+
+// addReflectOff adds a pointer to the reflection lookup map in the runtime.
+// It returns a new ID that can be used as a typeOff or textOff, and will
+// be resolved correctly. Implemented in the runtime package.
+func addReflectOff(ptr unsafe.Pointer) int32 {
+	i := len(typeOffList)
+	typeOffList = append(typeOffList, (*rtype)(ptr))
+	return int32(i)
+}
+
+func internalStr(strObj *js.Object) string {
+	var c struct{ str string }
+	js.InternalObject(c).Set("str", strObj) // get string without internalizing
+	return c.str
+}
+
+func isWrapped(typ Type) bool {
+	return jsType(typ).Get("wrapped").Bool()
+}
+
+func copyStruct(dst, src *js.Object, typ Type) {
+	fields := jsType(typ).Get("fields")
+	for i := 0; i < fields.Length(); i++ {
+		prop := fields.Index(i).Get("prop").String()
+		dst.Set(prop, src.Get(prop))
+	}
+}
+
+func makeValue(t Type, v *js.Object, fl flag) Value {
+	rt := t.common()
+	if t.Kind() == Array || t.Kind() == Struct || t.Kind() == Ptr {
+		return Value{rt, unsafe.Pointer(v.Unsafe()), fl | flag(t.Kind())}
+	}
+	return Value{rt, unsafe.Pointer(js.Global.Call("$newDataPointer", v, jsType(rt.ptrTo())).Unsafe()), fl | flag(t.Kind()) | flagIndir}
+}
+
+func MakeSlice(typ Type, len, cap int) Value {
+	if typ.Kind() != Slice {
+		panic("reflect.MakeSlice of non-slice type")
+	}
+	if len < 0 {
+		panic("reflect.MakeSlice: negative len")
+	}
+	if cap < 0 {
+		panic("reflect.MakeSlice: negative cap")
+	}
+	if len > cap {
+		panic("reflect.MakeSlice: len > cap")
+	}
+
+	return makeValue(typ, js.Global.Call("$makeSlice", jsType(typ), len, cap, js.InternalObject(func() *js.Object { return jsType(typ.Elem()).Call("zero") })), 0)
+}
+
+func TypeOf(i interface{}) Type {
+	if !initialized { // avoid error of uint8Type
```

Comment:
When does this happen?

Reply:
Unfortunately, I do not know. This is basically a verbatim citation of the code in `reflect` natives, and it seems to work (judging by the fact that it passes tests).

This code is 7 years old and I fully expect that it could be improved significantly. All that "avoid dead code elimination" magic, for example, is highly suspicious to me. According to my understanding of the GopherJS compiler it should not be necessary, but it exists for some reason. That said, I help myself back from making any substantial refactorings right now, since it is a risky endeavor and I didn't want to mix it with the Go 1.16 support work.